auth/src/lib.rs

```diff
@@ -0,0 +1,188 @@
+
+/// Provides marjor functions that one will need.
+///
+/// This is the primary type for using the `auth` module.
+/// It will interact with the database, generate and validate records.
+#[derive(Clone)]
+pub struct AuthClient {
+    collection: Collection<PermissionRecord>,
+    argon: Argon2<'static>,
+}
+
+impl AuthClient {
+    /// Create a new [`AuthClient`] with the given [`Collection`].
+    pub fn new(collection: Collection<PermissionRecord>) -> AuthClient {
+        AuthClient {
+            collection,
+            argon: Default::default(),
+        }
+    }
+
+    /// Get the inner [`Collection`].
+    pub fn collection(&self) -> Collection<PermissionRecord> {
+        self.collection.clone()
+    }
+
+    /// List all records in the database.
+    pub async fn list(&self) -> Result<Cursor<PermissionRecord>> {
+        self.collection.find(None, None).await.map_err(Into::into)
+    }
+
+    /// Return the count of records in the database.
+    ///
+    /// In debug mode, this will do a more expensive but accurate [`count_documents`].
+    /// In release mode, this will use [`estimated_document_count`],
+    /// which returns metadata of the database without actually call a `find`.
+    ///
+    /// For more detail, see document of [`count`] and [`countDocuments`] of mongodb.
+    ///
+    /// [`count_documents`]: Collection::count_documents
+    /// [`estimated_document_count`]: Collection::estimated_document_count
+    /// [`count`]: https://www.mongodb.com/docs/manual/reference/method/db.collection.count/
+    /// [`countDocuments`]: https://www.mongodb.com/docs/manual/reference/method/db.collection.countDocuments/
+    pub async fn count(&self) -> Result<u64> {
+        if cfg!(debug_assertions) {
+            self.collection.count_documents(None, None).await
+        } else {
+            self.collection.estimated_document_count(None).await
+        }
+        .map_err(Into::into)
+    }
+
+    /// Try insert a new record.
+    ///
+    /// Return wether the record is inserted.
+    /// If one record with same username exists, this will leave it intact.
+    pub async fn new_record(
+        &self,
+        username: impl Into<String>,
+        password: impl AsRef<[u8]>,
+        permission: PermissionSet,
+    ) -> Result<bool> {
+        let username = username.into();
+        let salt = SaltString::generate(&mut OsRng);
+        let hash = self.argon.hash_password(password.as_ref(), &salt)?;
+
+        let record = PermissionRecord::new(hash, username, permission);
+
+        let doc = to_bson(&record)?;
+        let res = self
+            .collection
+            .update_one(
+                doc! {
+                  "username" : record.username()
+                },
+                doc! {
+                 "$setOnInsert": doc
+                },
+                UpdateOptions::builder().upsert(true).build(),
+            )
+            .await?;
+
+        Ok(res.upserted_id.is_some())
+    }
```

Comment:
This function can only attach permissions to users during creation. So how do we modify permissions of a user?